screwdriver/api/api.go

```diff
@@ -33,102 +38,199 @@ type ResponseError struct {
+
 	if err != nil {
-		return nil, fmt.Errorf("Failed to get command from Screwdriver API: %v", err)
+		return nil, fmt.Errorf("Failed to get http response: %v", err)
 	}
-	defer res.Body.Close()
-	body, err := handleResponse(res)
+
+	// Check response body
+	bodyBytes, err = handleResponse(bodyBytes, statusCode)
 
 	if err != nil {
 		return nil, err
 	}
 
-	err = json.Unmarshal(body, cmd)
+	cmd := new(util.CommandSpec)
+	err = json.Unmarshal(bodyBytes, cmd)
 	if err != nil {
 		return nil, fmt.Errorf("Failed to unmarshal command: %v", err)
 	}
 	return cmd, nil
 }
+
+func specToPayloadBytes(commandSpec *util.CommandSpec) (specBytes []byte, err error) {
+	// Generate yaml payload bytes
+	commandSpecBytes, err := json.Marshal(&commandSpec)
+	var payload util.PayloadYaml
+	payload.Yaml = string(commandSpecBytes)
+	specBytes, err = json.Marshal(&payload)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+func (c client) PostCommand(specPath string, commandSpec *util.CommandSpec) (string, error) {
+	// Generate URL
+	uri, err := parseURL(c.baseURL)
+	if err != nil {
+		return "", fmt.Errorf("Failed to parse URL on POST: %v", err)
+	}
+	uri.Path = path.Join(uri.Path, "commands")
+
+	// Generate payload
+
+	// Prepare body
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+
+	// Convert command spec to payload
+	specBytes, err := specToPayloadBytes(commandSpec)
+	if err != nil {
+		return "", fmt.Errorf("Failed to convert spec to bytes:%v", err)
+	}
+
+	// Set yaml filename
+	fileNameYaml := filepath.Base(specPath)
+	yamlPart, err := writer.CreateFormFile("file", fileNameYaml)
+	if err != nil {
+		return "", fmt.Errorf("Failed to load spec file:%v", err)
+	}
+
+	// Write yaml part
+	yamlPart.Write(specBytes)
+
+	if commandSpec.Format == "binary" {
+		// Load binary from file
+		fileContentsBin, err := util.LoadByte(commandSpec.Binary.File)
+		if err != nil {
+			return "", fmt.Errorf("Failed to load binary file:%v", err)
+		}
+
+		// Set filename of binary
+		fileNameBin := filepath.Base(commandSpec.Binary.File)
+		binPart, err := writer.CreateFormFile("file", fileNameBin)
+		if err != nil {
+			return "", fmt.Errorf("Failed to create form of binary:%v", err)
+		}
+
+		// Write binary part
+		binPart.Write(fileContentsBin)
+	}
+	writer.Close()
+
+	// Send request
+	contentType := writer.FormDataContentType()
+	responseBytes, statusCode, err := c.httpRequest("POST", uri.String(), c.jwt, contentType, body)
+	if err != nil {
+		return "", fmt.Errorf("Post request failed: %v", err)
+	}
+
+	// Check response
+	responseBytes, err = handleResponse(responseBytes, statusCode)
+	if err != nil {
+		return "", err
+	}
+
+	// Get version of posted command
+	responseSpec := util.CommandSpec{}
+	err = json.Unmarshal(responseBytes, &responseSpec)
+	if err != nil {
+		return "", fmt.Errorf("Failed to parse json response %v", err)
+	}
+
+	return responseSpec.Version, nil
```

Comment:
I think to return util.CommandSpec is better, because this function's role is to post command data and get the response from api. I think deciding return parameter is not this function's role.

Reply:
Thank you! I've fixed.